xline-client/src/clients/lock.rs

```diff
@@ -16,26 +37,343 @@ pub struct LockClient {
+    /// If `CurpClient` fails to send request
+    #[inline]
+    pub async fn lock(&self, request: LockRequest) -> Result<LockResponse> {
+        let mut lease_id = request.inner.lease;
+        if lease_id == 0 {
+            let resp = self
+                .lease_client
+                .grant(LeaseGrantRequest::new(request.ttl))
+                .await?;
+            lease_id = resp.id;
+        }
+        let prefix = format!(
+            "{}/",
+            String::from_utf8_lossy(&request.inner.name).into_owned()
+        );
+        let key = format!("{prefix}{lease_id:x}");
+        let lock_success = AtomicBool::new(false);
+        let lock_inner = self.lock_inner(prefix, key.clone(), lease_id, &lock_success);
+        tokio::pin!(lock_inner);
+
+        LockFuture {
+            key,
+            lock_success: &lock_success,
+            lock_client: self,
+            lock_inner,
+        }
+        .await
+    }
+
+    /// The inner lock logic
+    async fn lock_inner(
+        &self,
+        prefix: String,
+        key: String,
+        lease_id: i64,
+        lock_success: &AtomicBool,
+    ) -> Result<LockResponse> {
+        let txn = Self::create_acquire_txn(&prefix, lease_id);
+        let (cmd_res, sync_res) = self.propose(txn, false).await?;
+        let mut txn_res = Into::<TxnResponse>::into(cmd_res.decode());
+        let my_rev = sync_res
+            .unwrap_or_else(|| unreachable!("sync_res always has value when use slow path"))
+            .revision();
+        let owner_res = txn_res
+            .responses
+            .swap_remove(1)
+            .response
+            .and_then(|r| {
+                if let Response::ResponseRange(res) = r {
+                    Some(res)
+                } else {
+                    None
+                }
+            })
+            .unwrap_or_else(|| unreachable!("owner_resp should be a Get response"));
+
+        let owner_key = owner_res.kvs;
+        let header = if owner_key
+            .get(0)
+            .map_or(false, |kv| kv.create_revision == my_rev)
+        {
+            owner_res.header
+        } else {
+            self.wait_delete(prefix, my_rev).await?;
+            let range_req = RangeRequest {
+                key: key.as_bytes().to_vec(),
+                ..Default::default()
+            };
+            let result = self.propose(range_req, true).await;
+            match result {
+                Ok(res) => {
+                    let res = Into::<RangeResponse>::into(res.0.decode());
+                    if res.kvs.is_empty() {
+                        return Err(ClientError::RpcError(String::from("session expired")));
+                    }
+                    res.header
+                }
+                Err(e) => {
+                    let _ignore = self.delete_key(key.as_bytes()).await;
+                    return Err(e);
+                }
+            }
+        };
+
+        let _ignore = lock_success.fetch_or(true, Ordering::Relaxed);
```

Comment:
Why do you invoke `fetch_or` here? We can directly store it as true.

Reply:
fixed